synapse/handlers/federation_event.py

```diff
@@ -953,6 +953,47 @@ async def _process_pulled_event(
             )
             return
 
+        # Check if we've already tried to process this event at some point in the past.
+        # We aren't concerned with the expontntial backoff here, just whether it has
+        # failed before.
+        failed_pull_attempt_info = await self._store.get_event_failed_pull_attempt_info(
+            event.room_id, event_id
+        )
+        if failed_pull_attempt_info:
+            # Process previously failed backfill events in the background to not waste
+            # time on something that is bound to fail again.
+            #
+            # TODO: Are we concerned with processing too many events in parallel since
+            # we just fire and forget this off to the background? Should we instead have
+            # a background queue to chew through?
+            run_as_background_process(
+                "_try_process_pulled_event",
+                self._try_process_pulled_event,
+                origin,
+                event,
+                backfilled,
+            )
```

Comment:
One test we may be able to add is a sanity check that events are still processed even if they have a failing attempt. As long as there is a way to wait for any background jobs. Something similar to `wait_for_background_updates()` but for background processes. Maybe we would have to return the `run_as_background_process(...)` deferred and use `wait_on_thread(deffered)`.

Better way to do this? Is it worth returning the deferred just so we can sanity check test?

Reply:
How do I know how much to `pump()` to gurantee that the background process had a chance to run? I've [used it in the tracing tests before](https://github.com/matrix-org/synapse/blob/ca5c4be92166775ec1de9e79a04db1e136609a1f/tests/logging/test_opentracing.py#L254-L279) but it was just magic I copied from somewhere else

Reply:
Calling it at all should do it -- you're giving up control to the reactor, so the reactor ticks over and does *stuff*, `pump(0.1)` is likely fine, but I'm not really sure.

Reply:
Added a sanity check test :white_check_mark:

Reply:
Conversation continued in https://github.com/matrix-org/synapse/pull/15585#discussion_r1204766827 (diff of the actual test)